### PR TITLE
feat: delete datasets in public-access s3 when tombstoning a Collection

### PR DESCRIPTION
**Pull request by danieljhegeman**
- fix error for submissions bucket info returned by ../s3-upload-credentials endpoint
- add datasets bucket env var to 'service' module

## Reason for Change

- Addresses first part of #4537 
- We will address cleaning up tombstoned datasets from public bucket by performing a mirror from prod->staging, and then staging->prod.

## Testing steps

- unit tests updated
- not yet rdev-tested

## Notes for Reviewer

This PR also needs to include invalidation of cloudfront caches for the datasets distribution. This work is in-progress and will be added soon.

### CODE REVIEW

File: backend/curation/api/v1/curation/collections/collection_id/s3_upload_credentials.py [hunk 1]
```diff
@@ -29,6 +30,9 @@ def get(collection_id: str, token_info: dict):
         upload_key_prefix = f"super/{collection_id}/"
     else:
         upload_key_prefix = f"{user_info.user_id}/{collection_id}/"
+    rdev_prefix = os.environ.get("REMOTE_DEV_PREFIX", "").strip("/")
+    if rdev_prefix:
+        upload_key_prefix = f"{rdev_prefix}/{upload_key_prefix}"
```
Comment (danieljhegeman) [inline backend/curation/api/v1/curation/collections/collection_id/s3_upload_credentials.py:35]: see comment above. Remote dev prefix is part of the _key_, not the _bucket_.

File: tests/unit/backend/layers/api/test_curation_api.py [hunk 2]
```diff
@@ -83,13 +83,21 @@ def test__delete_private_collection_by_collection_version_id(self):
...
+    def test__delete_tombstone_collection(self, mock_delete: Mock):
         tests = [("not_owner", 403), ("noauth", 401), ("owner", 403), ("super", 403)]
         for auth, expected_response in tests:
             with self.subTest(auth):
-                collection_id = self.generate_published_collection().collection_id
-                self.business_logic.tombstone_collection(collection_id)
-                self._test(collection_id, auth, expected_response)
+                collection = self.generate_published_collection()
+                self.business_logic.tombstone_collection(collection.collection_id)
+                if self.run_as_integration:
```
Comment (danieljhegeman) [inline tests/unit/backend/layers/api/test_curation_api.py:93]: no need to complicate the db mock code to handle this, so we'll just run the arg check for integration tests only

File: tests/unit/processing/base_processing_test.py [hunk 3]
```diff
@@ -22,6 +23,9 @@ def upload_file(self, src_file: str, bucket_name: str, dst_file: str, extra_args
     def upload_directory(self, src_dir: str, s3_uri: str):
         self.mock_s3_fs.append(s3_uri)
 
+    def delete_files(self, bucket_name: str, object_keys: List[str]):
+        pass
+
```
Comment (danieljhegeman) [inline tests/unit/processing/base_processing_test.py:28]: for completeness

File: tests/unit/backend/layers/api/test_curation_api.py [hunk 4]
```diff
@@ -83,13 +83,21 @@ def test__delete_private_collection_by_collection_version_id(self):
...
+                    dataset_version_ids = [d_v.version_id.id for d_v in collection.datasets]
+                    delete_args = set(mock_delete.call_args.args[1])
+                    expected_delete_args = set()
+                    for file_type in ("h5ad", "rds"):
+                        expected_delete_args.update([f"{d_v_id}.{file_type}" for d_v_id in dataset_version_ids])
```
Comment (danieljhegeman) [inline tests/unit/backend/layers/api/test_curation_api.py:98]: I will be doing the dataset file structure migration before this hits prod followed by mirrors from prod to dev and staging, so this will work (i.e. every dataset file will, at that point, live at a `<dataset_version_id>.<file_type>` key)

File: backend/layers/thirdparty/s3_provider.py [hunk 5]
```diff
@@ -48,6 +48,12 @@ def upload_file(self, src_file: str, bucket_name: str, dst_file: str, extra_args
             ExtraArgs=extra_args,
         )
 
+    def delete_files(self, bucket_name: str, object_keys: List[str]):
+        """
+        Deletes the objects `object_keys` from bucket `bucket_name`
+        """
+        self.client.delete_objects(Bucket=bucket_name, Delete={"Objects": [{"Key": key} for key in object_keys]})
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/layers/thirdparty/s3_provider.py:55]: Also add unit tests for the above cases.

File: tests/unit/backend/layers/business/test_business.py [hunk 6]
```diff
@@ -1475,5 +1475,32 @@ def test_dataset_published_at_with_replaced_dataset_ok(self):
         self.assertEqual(dataset_1.canonical_dataset.published_at, published_collection.published_at)
 
 
+class TestCollectionUtilities(BaseBusinessLogicTestCase):
+    def test__delete_datasets_from_public_access_bucket(self):
+        published_collection = self.initialize_published_collection()
+        new_version = self.business_logic.create_collection_version(published_collection.collection_id)
+
+        # We will replace the first dataset
```
Comment (Bento007) [inline tests/unit/backend/layers/business/test_business.py:1483]:
```suggestion
        # Update the first dataset
```

File: tests/unit/backend/layers/business/test_business.py [hunk 7]
```diff
@@ -1475,5 +1475,32 @@ def test_dataset_published_at_with_replaced_dataset_ok(self):
...
+        replaced_dataset_version_id, _ = self.business_logic.ingest_dataset(
+            new_version.version_id, "http://fake.url", None, dataset_id_to_replace
+        )
+
+        self.complete_dataset_processing_with_success(replaced_dataset_version_id)
+
+        dataset_version_ids = [d_v.version_id.id for d_v in published_collection.datasets] + [
+            replaced_dataset_version_id
+        ]
+        delete_keys = set(
```
Comment (Bento007) [inline tests/unit/backend/layers/business/test_business.py:1495]:
```suggestion
        actual_delete_keys = set(
```

File: tests/unit/backend/layers/business/test_business.py [hunk 8]
```diff
@@ -1475,5 +1475,32 @@ def test_dataset_published_at_with_replaced_dataset_ok(self):
         self.assertEqual(dataset_1.canonical_dataset.published_at, published_collection.published_at)
 
 
+class TestCollectionUtilities(BaseBusinessLogicTestCase):
+    def test__delete_datasets_from_public_access_bucket(self):
```
Comment (Bento007) [inline tests/unit/backend/layers/business/test_business.py:1479]: add a docstring. You are testing delete published datasets and unpublished datasets, add those details here.

File: tests/unit/backend/layers/business/test_business.py [hunk 9]
```diff
@@ -1475,5 +1475,32 @@ def test_dataset_published_at_with_replaced_dataset_ok(self):
         self.assertEqual(dataset_1.canonical_dataset.published_at, published_collection.published_at)
 
 
+class TestCollectionUtilities(BaseBusinessLogicTestCase):
```
Comment (Bento007) [inline tests/unit/backend/layers/business/test_business.py:1478]: Add a test where multiple collection versions refers to the same dataset version.

**Review by Bento007 — CHANGES_REQUESTED** [review]
A few edge cases to take into account

File: backend/curation/api/v1/curation/collections/collection_id/s3_upload_credentials.py (hunk 1 above)
Comment (danieljhegeman) [reply backend/curation/api/v1/curation/collections/collection_id/s3_upload_credentials.py:35]: It was not noticed. It would cause an issue only in rdev, and only in certain scenarios. But it was wrong and should be fixed.

File: backend/layers/business/business.py [hunk 10]
```diff
@@ -570,10 +571,29 @@ def delete_collection_version(self, version_id: CollectionVersionId) -> None:
...
         self.database_provider.delete_collection_version(version_id)
 
+    def delete_datasets_from_public_access_bucket(self, collection_id: CollectionId) -> List[str]:
+        """
+        Delete all associated publicly-accessible Datasets in s3
+        """
+        collection_versions = self.database_provider.get_all_versions_for_collection(collection_id)
+        datasets_bucket = os.getenv("DATASETS_BUCKET")
+        rdev_prefix = os.environ.get("REMOTE_DEV_PREFIX", "").strip("/")
+        object_keys = []
```
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/business/business.py:581]: great point

File: tests/unit/backend/layers/business/test_business.py (hunk 9 above)
Comment (danieljhegeman) [reply tests/unit/backend/layers/business/test_business.py:1478]: that is encompassed in the existing test since one dataset is updated and the other is not.

File: backend/layers/thirdparty/s3_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: > We don't want to silently fail to delete. If we do fail we need to roll back the and database changes made. t's is an edge case, but how difficult are these datasets to track down if this fails silently?

In some cases I would agree with you, but there are some important considerations: the user is _trying_ to delete these assets; the public dataset deletion is occurring before the updates to entities in the database, so we guarantee that we don't accidentally leave datasets behind -- the request only fails silently if the key doesn't exist, which does not constitute a problem in this scenario. If the actual attempt to delete fails, perhaps for network reasons, the request fails loudly and the Collection will not be deleted from the db. The user sees 500. The user will therefore be aware that the Collection still exists. Perhaps we should capture 500s for deletes and inform them that public assets may be missing.

Actually, in lieu of that, we can perform a HEAD request for all asset files for all `dataset.assets[]` objects. That way we can say whether or not the public asset is still available. If it's not, then the curator can get in touch with us.

cc @nayib-jose-gloria for thoughts on this issue

[the remark this answers is not in the document]
Comment (Bento007) [reply backend/layers/thirdparty/s3_provider.py]: I changed my mind we should make the request without `quite` and log the response. Knowing what request deleted a file could help with debugging and audits in the future.

File: tests/unit/backend/layers/business/test_business.py (hunk 9 above)
Comment (Bento007) [reply tests/unit/backend/layers/business/test_business.py:1478]: I would mention that in the docstring, that test case is obfuscated in the `initialize_published_collection`

File: backend/layers/business/business.py [hunk 11]
```diff
@@ -570,10 +571,28 @@ def delete_collection_version(self, version_id: CollectionVersionId) -> None:
...
+        rdev_prefix = os.environ.get("REMOTE_DEV_PREFIX", "").strip("/")
+        object_keys = set()
+        for collection_version in collection_versions:
+            for d_v in collection_version.datasets:
+                for file_type in ("h5ad", "rds"):
+                    dataset_version_s3_object_key = f"{d_v.version_id.id}.{file_type}"
+                    if rdev_prefix:
+                        dataset_version_s3_object_key = f"{rdev_prefix}/{dataset_version_s3_object_key}"
+                    object_keys.add(dataset_version_s3_object_key)
+        self.s3_provider.delete_files(bucket, list(object_keys))
```
Comment (Bento007) [inline backend/layers/business/business.py:588]: do we need to convert it into a list?

File: backend/layers/thirdparty/s3_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply backend/layers/thirdparty/s3_provider.py]: I think capturing 500s and returning a message to the user that tombstoning failed, but some public assets may have already been deleted, and suggesting they try again is a reasonable route. Perhaps we should introduce our own retry (unless boto3 already has that as a default?) here to account for transient network blips

File: backend/layers/thirdparty/s3_provider.py [hunk 12]
```diff
@@ -48,6 +48,18 @@ def upload_file(self, src_file: str, bucket_name: str, dst_file: str, extra_args
             ExtraArgs=extra_args,
         )
 
+    def delete_files(self, bucket_name: str, object_keys: List[str]):
+        """
+        Deletes the objects `object_keys` from bucket `bucket_name`
+        """
+
+        def chunk_list(items: List[str], chunk_size: int = 1000):  # s3 client delete_objects accepts up to 1000 keys
```
Comment (Bento007) [inline backend/layers/thirdparty/s3_provider.py:56]: I don't think this is complicated enough to need sub function. Please inline.

File: backend/layers/thirdparty/s3_provider.py [hunk 13]
```diff
@@ -48,6 +48,18 @@ def upload_file(self, src_file: str, bucket_name: str, dst_file: str, extra_args
...
+            for i in range(0, len(items), chunk_size):
+                yield items[i : i + chunk_size]
+
+        for keys_batch in chunk_list(object_keys):
```
Comment (Bento007) [inline backend/layers/thirdparty/s3_provider.py:60]: We should make the 1000 into a constant.

```suggestion
        AWS_S3_BULK_DELETE_LIMIT = 1000
        for i in range(0, len(object_keys), AWS_S3_BULK_DELETE_MAX):
            keys_batch = object_keys[ i: i+AWS_S3_BULK_DELETE_MAX]
```

**Review by Bento007 — CHANGES_REQUESTED** [review]

File: backend/layers/thirdparty/s3_provider.py [hunk 14]
```diff
@@ -48,6 +48,18 @@ def upload_file(self, src_file: str, bucket_name: str, dst_file: str, extra_args
...
+            self.client.delete_objects(Bucket=bucket_name, Delete={"Objects": [{"Key": key} for key in keys_batch]})
```
Comment (Bento007) [inline backend/layers/thirdparty/s3_provider.py:61]: log this response

File: backend/layers/thirdparty/s3_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply backend/layers/thirdparty/s3_provider.py]: Agree w/ logging responses for debugging should it come up

[the remark this answers is not in the document]
Comment (brianraymor) [reply backend/layers/thirdparty/s3_provider.py]: > returning a message to the user that tombstoning failed, but some public assets may have already been deleted, and suggesting they try again is a reasonable route.

That sounds like ambiguity/complexity is being pushed to the user. The operation should either succeed or not?

[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply backend/layers/thirdparty/s3_provider.py]: There's always the rare possibility that a network error causes some object(s) not to be deleted in a bulk delete request. We can enable dataset-assets to be a versioned bucket, which would allow us to restore deleted objects and rollback if we come across an error--but that seems like overengineering for this edge case IMO 

We can alternatively bulk delete and then alert engineers of an error if the boto3 delete_objects command fails on any asset + return the list of objects it failed to delete, like we do with upload errors. And users can see 500 and follow the same workflow as with upload errors

[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: >That sounds like ambiguity/complexity is being pushed to the user. The operation should either succeed or not?

That is true, and in all likelihood we will never run into this problem. Really what we're dealing with here is the unlikely event that _some_ of the assets are deleted successfully and then a transient error causes the whole operation to return 500 (meaning the Collection entity does not actually get marked as tombstoned/deleted), at which point we'd be left with the Collection still in play and some of the assets gone, but not all. We don't have a way to perfectly transactionalize this operation since it is db + s3 without putting a lot of effort into it.

Plus, the ambiguity will only ever get put on the user _if_ they have tried to delete the Collection (which means they would surely just try to delete it again anyways). I don't want us to spend time solving perfectly the scenario where someone tries to delete something, it fails, and then they decide they don't want to delete it anymore. 😄

File: backend/layers/thirdparty/s3_provider.py [hunk 15]
```diff
@@ -52,13 +57,14 @@ def delete_files(self, bucket_name: str, object_keys: List[str]):
...
-        def chunk_list(items: List[str], chunk_size: int = 1000):  # s3 client delete_objects accepts up to 1000 keys
-            for i in range(0, len(items), chunk_size):
-                yield items[i : i + chunk_size]
-
-        for keys_batch in chunk_list(object_keys):
-            self.client.delete_objects(Bucket=bucket_name, Delete={"Objects": [{"Key": key} for key in keys_batch]})
+        for i in range(0, len(object_keys), AWS_S3_MAX_ITEMS_PER_BATCH):
+            key_batch = object_keys[i:AWS_S3_MAX_ITEMS_PER_BATCH]
+            logger.info(f"Deleting assets from public-access bucket '{bucket_name}': {key_batch}")
+            self.client.delete_objects(
```
Comment (Bento007) [inline backend/layers/thirdparty/s3_provider.py:63]: We should also log the Deleted and Errors fields from this response.

Comment (Bento007) [reply backend/layers/thirdparty/s3_provider.py:63]: The value mentioned above are hard to find after deleting the files. Since deletion is a sensitive topic, I'd rather error on the side of caution and log more than we might need.

File: backend/layers/thirdparty/s3_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: @Bento007 @nayib-jose-gloria  I actually think further optimization here is not worth the effort, and here's why: the curator only performs this action if they intend to erase the Collection. If the operation fails, i.e., if any of the deletes don't succeed, then the curator will be made aware of this in the UI because the request will fail. Therefore they've been made aware that the Collection still exists. They'll either try again, and this code will succeed or fail (try, try again), and then they'll contact us if it continues failing. When they contact us, our move will be to delete everything, which is extremely easy to do with a couple postgres queries for dataset version ids; it will not even be worth referencing the logs to figure out exactly what happened with the failed delete attempts. cc @brianraymor

[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: So I submit that we leave it with the current level of logging. What really matters is having the s3 deletes coming _before_ the tombstoning of the Collection entity, which is what this code is doing.

[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply backend/layers/thirdparty/s3_provider.py]: Actually, Dan, are you sure it won't fail silently for all errors? Per the [boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/delete_objects.html), it seems like it returns the same object response shape regardless, which provides list of Errors. So I think we'd need to check here for `if resp['Errors']` and raise an exception if list has entries

**Review by nayib-jose-gloria — CHANGES_REQUESTED** [review]
1 more note about error handling, otherwise looking good

File: backend/layers/thirdparty/s3_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: It fails loudly if the request is malformed, if permissions are not correct, etc.

So if the key doesn't exist, it fails silently. But that would indicate a business logic failing... and our standard for interactions with s3 so far has not been to double check the key name for, say, writes.

[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: I can't imagine a network failure would be swallowed by the boto client; that would surely be a loud failure as well.

[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply backend/layers/thirdparty/s3_provider.py]: It 'fails loudly' by returning the error in the Errors list though, not by throwing an Exception. And if we don't check the Errors list for entries, aren't we failing silently as a result?

**Review by Bento007 — APPROVED** [review]

File: backend/layers/thirdparty/s3_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/layers/thirdparty/s3_provider.py]: Fail loudly == throws Exception
